lib/test_ids/allocator.rb

```diff
@@ -109,6 +166,11 @@ def allocate(instance, options)
         number['size'] ||= number_size
         softbin['number'] ||= allocate_softbin(bin: bin['number'], number: number['number'], size: softbin_size)
         softbin['size'] ||= softbin_size
+      elsif options[:softbin].is_a?(Range)
+        softbin['number'] ||= allocate_softbin(bin: bin['number'], softbin: options[:softbin], size: softbin_size)
```

Comment:
I'll attempt to explain.

in this code, the else condition is the one that gets executed because
~~~
config.softbins.algorithm && config.softbins.algorithm.to_s =~ /n/ = nil 
(config.softbins.callback && !config.numbers.function?) = false 
~~~

With the code statement below

~~~
softbin['number'] ||= allocate_softbin(bin: bin['number'], size: softbin_size)
~~~

We are just passing the bin and softbin size as options to the allocate_softbin method, which in turn calls the callback.

~~~
callback.call(bin, options)
~~~

The options at this time contain only the bin and size that we passed in before.

Passing just the options to allocate_softbin will not work either because size is used in the range_item method to increment the pointer by the size configured for either bin/softbin/number.

~~~
From range_item method:

       if previous_assigned_value == range.to_a[@pointer]
          @pointer += options[:size]
          assigned_value = range.to_a[@pointer]
        else
          # Because of the pointer calculations above, I don't think it will ever reach here, has not in my test cases so far!
          assigned_value = range.to_a[@pointer]
        end
~~~

Reply:
This should not be required, but I guess it is working around the fact that the application is passing in something like options[:softbin] == (1000..2999) and this code is trying to work around the existing convention that any softbin option passed in here will take precedence and it will not fall through the to callback.

This should be the application's responsibility to resolve, so actually your example code should really be:

~~~ruby
  config.softbins do |bin, options|
    if options[:softbin_range] 
      TestIds.next_in_range(options[:softbin_range])
    end
  end
~~~

i.e. it should not used the reserved option :softbin to pass the range about.